<a href="https://colab.research.google.com/github/mayujie/MSC_text_mining/blob/master/RD_Embedding.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# [Amazon product data](http://jmcauley.ucsd.edu/data/amazon/index_2014.html)
# [Amazon product data Julian McAuley, UCSD](http://jmcauley.ucsd.edu/data/amazon/)
Files

"Small" subsets for experimentation

https://zhuanlan.zhihu.com/p/137537499

http://baijiahao.baidu.com/s?id=1665370803562402243

http://www.xiaomaidong.com/?p=1249

https://towardsdatascience.com/double-your-google-colab-ram-in-10-seconds-using-these-10-characters-efa636e646ff


In [ ]:
from google.colab import drive
drive.mount('/content/drive/')

Mounted at /content/drive/


In [ ]:
root_path = '/content/drive/My Drive/Colab/MSC_RD'
root_path

'/content/drive/My Drive/Colab/MSC_RD'

## **Read from cleaned text file**

In [ ]:
# ### download text data
# !gdown https://drive.google.com/uc?id=1PtUMY6KuShzR2lk_85VbNizGT_wciBNl
# !unzip reviews_text.zip
# !rm /content/reviews_text.zip

In [ ]:
# handle_list = []
# with open('/content/reviews_text.txt', 'r') as file:
#   for line in file:
#     handle_list.append(line)

In [ ]:
# text_ls = map(lambda s: s.strip(), handle_list)
# text_ls = list(text_ls)
# text_ls[:5]

In [ ]:
### download cleaned text data
!gdown https://drive.google.com/uc?id=1Kb8ApeUYUDMOO39utV-dQ00PZ0s8yOAb 
!unzip cleaned_reviews_text.zip
!rm /content/cleaned_reviews_text.zip
!ls -lh *.txt

Downloading...
From: https://drive.google.com/uc?id=1Kb8ApeUYUDMOO39utV-dQ00PZ0s8yOAb
To: /content/cleaned_reviews_text.zip
76.6MB [00:00, 164MB/s] 
Archive:  cleaned_reviews_text.zip
  inflating: cleaned_reviews_text.txt  
-rw-r--r-- 1 root root 241M Nov 13 13:20 cleaned_reviews_text.txt


In [ ]:
cleaned_text_ls = []
with open('/content/cleaned_reviews_text.txt', 'r') as file:
  for line in file:
    cleaned_text_ls.append(line)

In [ ]:
print(len(cleaned_text_ls))
# text_ls[:5]
cleaned_text_ls[:5]

484826


['big expectations love english tv <COMMA> particular investigative detective stuff guy really boring <PERIOD> not appeal <PERIOD>\n',
 'highly recommend series <PERIOD> must anyone yearning watch grown television <PERIOD> complex characters plots keep one totally involved <PERIOD> thank amazin prime <PERIOD>\n',
 'one real snoozer <PERIOD> not believe anything read hear <COMMA> awful <PERIOD> no idea title means <PERIOD> neither <PERIOD>\n',
 'mysteries interesting <PERIOD> tension robson tall blond good not always believable <PERIOD> often seemed uncomfortable <PERIOD>\n',
 'show always excellent <COMMA> far british crime mystery one best ever made <PERIOD> stories well done acting top notch interesting twists realistic brutal storylines <PERIOD> show pulls no punches enters twisted minds criminals profiler psychiatrist helps northern english city police force <PERIOD> show looks like shot manchester called another name show <PERIOD> one episode not disc excellent prayer bone seperat

## remove newline '\n' from each row

In [ ]:
cleaned_text_ls = map(lambda s: s.strip('\n'), cleaned_text_ls)
cleaned_text_ls = list(cleaned_text_ls)
cleaned_text_ls[:5]

['big expectations love english tv <COMMA> particular investigative detective stuff guy really boring <PERIOD> not appeal <PERIOD>',
 'highly recommend series <PERIOD> must anyone yearning watch grown television <PERIOD> complex characters plots keep one totally involved <PERIOD> thank amazin prime <PERIOD>',
 'one real snoozer <PERIOD> not believe anything read hear <COMMA> awful <PERIOD> no idea title means <PERIOD> neither <PERIOD>',
 'mysteries interesting <PERIOD> tension robson tall blond good not always believable <PERIOD> often seemed uncomfortable <PERIOD>',
 'show always excellent <COMMA> far british crime mystery one best ever made <PERIOD> stories well done acting top notch interesting twists realistic brutal storylines <PERIOD> show pulls no punches enters twisted minds criminals profiler psychiatrist helps northern english city police force <PERIOD> show looks like shot manchester called another name show <PERIOD> one episode not disc excellent prayer bone seperate disc <

In [ ]:
# ## read data into dataframe
# import pandas as pd
# df_review = pd.DataFrame(cleaned_text_ls) 
# df_review

In [ ]:
# text_ls[2360363]
# cleaned_text_ls[2360363]

In [ ]:
# text_ls[9568]
cleaned_text_ls[9568]

'man phones lawyer asks <COMMA> much would charge just answering three simple questions <QUESTION> lawyer replies <COMMA> thousand dollars <PERIOD> thousand dollars <EXCLAMATION> exclaims man <PERIOD> expensive not <QUESTION> certainly <COMMA> says lawyer <PERIOD> <COMMA> third question <QUESTION> one <PERIOD> three lawyers buried necks cement <QUESTION> not enough cement <PERIOD> usa network keeps not effing <PERIOD> far concerned <COMMA> suits newest addictive show television <COMMA> not normally drawn legal dramas <PERIOD> witty funny smart loaded patented easy breezy sense fun usa network nowadays specializes <PERIOD> premise irresistible <COMMA> cast irrepressible <PERIOD> two leads <COMMA> gabriel macht patrick <PERIOD> adams <COMMA> so good <PERIOD> not not rave <PERIOD> suits chronicles exploits prestigious new york corporate law firm pearson hardman <PERIOD> pointedly <COMMA> catchy premise revolves around sleek newly made senior partner harvey specter macht dips shady shenani

In [ ]:
# text_ls[511526], text_ls[1639671], text_ls[1821690]
# cleaned_text_ls[511526], cleaned_text_ls[1639671], cleaned_text_ls[1821690]

In [ ]:
# def find_noise(reg, datalist):
#   count = 0
#   found_list = []

#   for i, row in enumerate(datalist):
#     found = re.findall(reg, row)

#     if found != []:
#       count += 1
#       # print(found[0])
#       # found = list(filter(None, found[0]))
#       # print(found)
#       found_list.append([i ,found])
  
#   return count, found_list
#   # return found_list

In [ ]:
# import re
# ## check tokens for each row
# count_key, found_key = find_noise(r"<\w+>", cleaned_text_ls)
# print(count_key, len(found_key))
# found_key[:10]

In [ ]:
# ss = []
# for i in found_key:
#   for j in i[1]:
#     ss.append(j)

# from collections import Counter 
# print(len(Counter(ss).most_common()))
# Counter(ss).most_common() # 12, 14 tokens with each amount

In [ ]:
cleaned_text_ls[:10]

['big expectations love english tv <COMMA> particular investigative detective stuff guy really boring <PERIOD> not appeal <PERIOD>',
 'highly recommend series <PERIOD> must anyone yearning watch grown television <PERIOD> complex characters plots keep one totally involved <PERIOD> thank amazin prime <PERIOD>',
 'one real snoozer <PERIOD> not believe anything read hear <COMMA> awful <PERIOD> no idea title means <PERIOD> neither <PERIOD>',
 'mysteries interesting <PERIOD> tension robson tall blond good not always believable <PERIOD> often seemed uncomfortable <PERIOD>',
 'show always excellent <COMMA> far british crime mystery one best ever made <PERIOD> stories well done acting top notch interesting twists realistic brutal storylines <PERIOD> show pulls no punches enters twisted minds criminals profiler psychiatrist helps northern english city police force <PERIOD> show looks like shot manchester called another name show <PERIOD> one episode not disc excellent prayer bone seperate disc <

In [ ]:
# reviews_text_merge = " ".join(cleaned_text_ls)

In [ ]:
# print(len(reviews_text_merge))
# reviews_text_merge[:1000]

In [ ]:
# with open("reviews_text_merge.txt", "w") as file:
#     file.write("%s" % reviews_text_merge)

In [ ]:
# # Compress review text merge into zip file
# !zip "{root_path}"/reviews_text_merge.zip reviews_text_merge.txt

In [ ]:
# # https://drive.google.com/file/d/1Wrw1Wk2w6pkdfdQ3asOTX1_s_7djo1yk/view?usp=sharing
# !unzip "{root_path}"/reviews_text_merge.zip

In [ ]:
# with open('/content/reviews_text_merge.txt') as file:
#     text = file.read()

In [ ]:
# # words = text.split()
# words = reviews_text_merge.split()

In [ ]:
# from collections import Counter 

# word_counts = Counter(words)
# print('Total number of words in cleaned corpus:', len(words))
# print('Number of unique words in cleaned corpus:', len(Counter(words)), len(Counter(words).most_common()))

Total number of words in cleaned corpus: 36436399
Number of unique words in cleaned corpus: 55984 55984


Total number of words in cleaned corpus: 36436399

Number of unique words in cleaned corpus: 55984 55984

In [ ]:
# word_counts.most_common()[44187-10:]

In [ ]:
# vocab_to_int, int_to_vocab = create_lookup_tables(words)

In [ ]:
# int_words = [vocab_to_int[word] for word in words]

In [ ]:
# print(words[:30])
# print(int_words[:30])

['big', 'expectations', 'love', 'english', 'tv', '<COMMA>', 'particular', 'investigative', 'detective', 'stuff', 'guy', 'really', 'boring', '<PERIOD>', 'not', 'appeal', '<PERIOD>', 'highly', 'recommend', 'series', '<PERIOD>', 'must', 'anyone', 'yearning', 'watch', 'grown', 'television', '<PERIOD>', 'complex', 'characters']
[101, 1803, 20, 1865, 791, 1, 749, 19071, 4452, 169, 785, 16, 1481, 0, 2, 2068, 0, 198, 90, 318, 0, 226, 311, 8751, 409, 1978, 2657, 0, 1951, 577]


['big', 'expectations', 'love', 'english', 'tv', '<COMMA>', 'particular', 'investigative', 'detective', 'stuff', 'guy', 'really', 'boring', '<PERIOD>', 'not', 'appeal', '<PERIOD>', 'highly', 'recommend', 'series', '<PERIOD>', 'must', 'anyone', 'yearning', 'watch', 'grown', 'television', '<PERIOD>', 'complex', 'characters']

[101, 1803, 20, 1865, 791, 1, 749, 19071, 4452, 169, 785, 16, 1481, 0, 2, 2068, 0, 198, 90, 318, 0, 226, 311, 8751, 409, 1978, 2657, 0, 1951, 577]

## create word_counts dictionary for all words

In [ ]:
%whos

Variable          Type             Data/Info
--------------------------------------------
cleaned_text_ls   list             n=484826
drive             module           <module 'google.colab.dri<...>s/google/colab/drive.py'>
file              TextIOWrapper    <_io.TextIOWrapper name='<...>ode='r' encoding='UTF-8'>
line              str              nice head lamp <EXCLAMATI<...>flashing modes <PERIOD>\n
root_path         str              /content/drive/My Drive/Colab/MSC_RD


In [ ]:
words = []
for row in cleaned_text_ls:
  for i in row.split():
    # print(i)
    words.append(i)

In [ ]:
from collections import Counter 

print('Total number of words in cleaned corpus:', len(words))
print('Number of unique words in cleaned corpus:', len(Counter(words)), len(Counter(words).most_common()))

Total number of words in cleaned corpus: 36436399
Number of unique words in cleaned corpus: 55984 55984


### For 6 classes data:

Length of reviews: **199096**

Total number of words in cleaned corpus: **18241862**

Number of unique words in cleaned corpus: **44187** **44187**

### **For 8 classes data (Current):**
Length of reviews: **484826**

Total number of words in cleaned corpus: **36436399**

Number of unique words in cleaned corpus: **55984** **55984**

### For 10 classes data:

Length of reviews: **2360871**

Total number of words in cleaned corpus: **145318404**

Number of unique words in cleaned corpus: **110454 110454**

In [ ]:
# Remove all words with  5 or fewer occurences
from collections import Counter 

def create_lookup_tables(words):
  word_counts = Counter(words)
  # print(word_counts.most_common(10))
  # print(word_counts.most_common())
  ## sorting the words from most to least frequent in text occurrence, descending
  sorted_vocab = sorted(word_counts, key=word_counts.get, reverse=True)
  # print(len(sorted_vocab)) # 110454
  ## create int_to_vocab dictionaries
  int_to_vocab = {ii: word for ii, word in enumerate(sorted_vocab)}
  vocab_to_int = {word: ii for ii, word in int_to_vocab.items()}

  return vocab_to_int, int_to_vocab

[('\<PERIOD\>', 1671352),
 ('\<COMMA\>', 1266882),
 ('not', 341506),
 ('album', 140815),
 ('one', 140781),
 ('like', 127551),
 ('so', 120402),
 ('\<EXCLAMATION\>', 95972),
 ('just', 95320),
 ('good', 89656)]

12 Tokens for 6 classes data

[('\<PERIOD\>', 1671352),
 ('\<COMMA\>', 1266882),
 ('\<EXCLAMATION\>', 95972),
 ('\<QUESTION\>', 38148),
 ('\<PRICE\>', 9371),
 ('\<HUNDREDPERCENTAGE\>', 1098),
 ('\<NOTCONTRACT\>', 638),
 ('\<MENTION\>', 197),
 ('\<AMAZONURL\>', 185),
 ('\<URL\>', 136),
 ('\<UNK\>', 54),
 ('\<YOUTUBEURL\>', 8)]

 **14 Tokens for 8 classes data (Current)**

 [('\<PERIOD\>', 3476956),
 ('\<COMMA\>', 2399259),
 ('\<EXCLAMATION\>', 211162),
 ('\<QUESTION\>', 64627),
 ('\<PRICE\>', 29257),
 ('\<HUNDREDPERCENTAGE\>', 4389),
 ('\<NOTCONTRACT\>', 986),
 ('\<AMAZONURL\>', 506),
 ('\<MENTION\>', 288),
 ('\<URL\>', 246),
 ('\<UNK\>', 138),
 ('\<YOUTUBEURL\>', 25),
 ('\<GOOGLEURL\>', 13),
 ('\<STAR\>', 9)]

In [ ]:
vocab_to_int, int_to_vocab = create_lookup_tables(words)

In [ ]:
int_words = [vocab_to_int[word] for word in words]

In [ ]:
print(words[:30])
print(int_words[:30])

['big', 'expectations', 'love', 'english', 'tv', '<COMMA>', 'particular', 'investigative', 'detective', 'stuff', 'guy', 'really', 'boring', '<PERIOD>', 'not', 'appeal', '<PERIOD>', 'highly', 'recommend', 'series', '<PERIOD>', 'must', 'anyone', 'yearning', 'watch', 'grown', 'television', '<PERIOD>', 'complex', 'characters']
[101, 1803, 20, 1865, 791, 1, 749, 19071, 4452, 169, 785, 16, 1481, 0, 2, 2068, 0, 198, 90, 318, 0, 226, 311, 8751, 409, 1978, 2657, 0, 1951, 577]


In [ ]:
import numpy as np

In [ ]:
int_words = []
for row in cleaned_text_ls:
  int_words.append([vocab_to_int[word] for word in row.split()])

In [ ]:
print(cleaned_text_ls[0])
print(words[:30])
print(int_words[0])

big expectations love english tv <COMMA> particular investigative detective stuff guy really boring <PERIOD> not appeal <PERIOD>
['big', 'expectations', 'love', 'english', 'tv', '<COMMA>', 'particular', 'investigative', 'detective', 'stuff', 'guy', 'really', 'boring', '<PERIOD>', 'not', 'appeal', '<PERIOD>', 'highly', 'recommend', 'series', '<PERIOD>', 'must', 'anyone', 'yearning', 'watch', 'grown', 'television', '<PERIOD>', 'complex', 'characters']
[101, 1803, 20, 1865, 791, 1, 749, 19071, 4452, 169, 785, 16, 1481, 0, 2, 2068, 0]


## Batch of data

In [ ]:
%whos

Variable               Type             Data/Info
-------------------------------------------------
Counter                type             <class 'collections.Counter'>
cleaned_text_ls        list             n=484826
create_lookup_tables   function         <function create_lookup_tables at 0x7ff193942950>
drive                  module           <module 'google.colab.dri<...>s/google/colab/drive.py'>
file                   TextIOWrapper    <_io.TextIOWrapper name='<...>ode='r' encoding='UTF-8'>
i                      str              <PERIOD>
int_to_vocab           dict             n=55984
int_words              list             n=484826
line                   str              nice head lamp <EXCLAMATI<...>flashing modes <PERIOD>\n
root_path              str              /content/drive/My Drive/Colab/MSC_RD
row                    str              nice head lamp <EXCLAMATI<...>> flashing modes <PERIOD>
vocab_to_int           dict             n=55984
words                  list         

In [ ]:
# cleaned_text_ls = [row.split() for row in cleaned_text_ls]

In [ ]:
# print(cleaned_text_ls[:5])

In [ ]:
for i, row in enumerate(cleaned_text_ls):
  cleaned_text_ls[i] = [word for word in row.split()]

In [ ]:
print(cleaned_text_ls[:5])
print(int_words[:5])

[['big', 'expectations', 'love', 'english', 'tv', '<COMMA>', 'particular', 'investigative', 'detective', 'stuff', 'guy', 'really', 'boring', '<PERIOD>', 'not', 'appeal', '<PERIOD>'], ['highly', 'recommend', 'series', '<PERIOD>', 'must', 'anyone', 'yearning', 'watch', 'grown', 'television', '<PERIOD>', 'complex', 'characters', 'plots', 'keep', 'one', 'totally', 'involved', '<PERIOD>', 'thank', 'amazin', 'prime', '<PERIOD>'], ['one', 'real', 'snoozer', '<PERIOD>', 'not', 'believe', 'anything', 'read', 'hear', '<COMMA>', 'awful', '<PERIOD>', 'no', 'idea', 'title', 'means', '<PERIOD>', 'neither', '<PERIOD>'], ['mysteries', 'interesting', '<PERIOD>', 'tension', 'robson', 'tall', 'blond', 'good', 'not', 'always', 'believable', '<PERIOD>', 'often', 'seemed', 'uncomfortable', '<PERIOD>'], ['show', 'always', 'excellent', '<COMMA>', 'far', 'british', 'crime', 'mystery', 'one', 'best', 'ever', 'made', '<PERIOD>', 'stories', 'well', 'done', 'acting', 'top', 'notch', 'interesting', 'twists', 'reali

In [ ]:
print(cleaned_text_ls[0])
print(int_words[0])

['big', 'expectations', 'love', 'english', 'tv', '<COMMA>', 'particular', 'investigative', 'detective', 'stuff', 'guy', 'really', 'boring', '<PERIOD>', 'not', 'appeal', '<PERIOD>']
[101, 1803, 20, 1865, 791, 1, 749, 19071, 4452, 169, 785, 16, 1481, 0, 2, 2068, 0]


In [ ]:
%whos

Variable               Type             Data/Info
-------------------------------------------------
Counter                type             <class 'collections.Counter'>
cleaned_text_ls        list             n=484826
create_lookup_tables   function         <function create_lookup_tables at 0x7ff193942950>
drive                  module           <module 'google.colab.dri<...>s/google/colab/drive.py'>
file                   TextIOWrapper    <_io.TextIOWrapper name='<...>ode='r' encoding='UTF-8'>
i                      int              484825
int_to_vocab           dict             n=55984
int_words              list             n=484826
line                   str              nice head lamp <EXCLAMATI<...>flashing modes <PERIOD>\n
root_path              str              /content/drive/My Drive/Colab/MSC_RD
row                    str              nice head lamp <EXCLAMATI<...>> flashing modes <PERIOD>
vocab_to_int           dict             n=55984
words                  list           